### PR TITLE
CRIMAPP-721 warn about incomplete records on CYA and submission review

### PR DESCRIPTION
**Pull request by timpeat**
## Description of change
Add validation to CYA and final review to indicate where there are incomplete sections. 

## Link to relevant ticket
[CRIMAPP-721](https://dsdmoj.atlassian.net/browse/CRIMAPP-721)
[CRIMAPP-717](https://dsdmoj.atlassian.net/browse/CRIMAPP-717)

## Notes for reviewer

NOTE: this change is as designed. The validation shows before save, and is also shown in a different format to other validations / warnings.

Extracts some of the means logic into a TypeOfMeansAssessment concern, so that it can be used in validations and decision tree, and also the forthcoming task list work. Not mad about this pattern, but we can see if it works when the task list logic added.

### After changes:

Screenshot 2024-04-17 at 10 02 43

Screenshot 2024-04-17 at 09 57 07

## How to manually test the feature

[CRIMAPP-721]: https://dsdmoj.atlassian.net/browse/CRIMAPP-721?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CRIMAPP-717]: https://dsdmoj.atlassian.net/browse/CRIMAPP-717?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/views/steps/outgoings/answers/edit.html.erb [hunk 1]
```diff
@@ -2,7 +2,7 @@
 <% step_header %>
 
 <div class="govuk-grid-row">
-  <div class="govuk-grid-column-two-thirds">
```
Comment (timpeat) [inline app/views/steps/outgoings/answers/edit.html.erb:5]: fixes CRIMAPP-717

File: app/validators/capital_assessment/answers_validator.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      record.errors.add :base, :incomplete_records if any_incomplete_records?(record)
+    end
+
+    private
+
+    def any_incomplete_records?(capital)
+      incomplete?(capital.properties) ||
+        incomplete?(capital.savings) ||
+        incomplete?(capital.investments) ||
+        incomplete?(capital.national_savings_certificates) ||
```
Comment (hiboabd) [inline app/validators/capital_assessment/answers_validator.rb:13]: Duplication of national savings certificate

File: app/presenters/summary/sections/case_details.rb [hunk 3]
```diff
@@ -98,9 +98,7 @@ def answers
 
       private
 
-      def kase
-        @kase ||= crime_application.case
-      end
+      delegate :kase, to: :crime_application
```
Comment (timpeat) [inline app/presenters/summary/sections/case_details.rb:101]: not needed as in base class now

File: app/services/decisions/income_decision_tree.rb [hunk 4]
```diff
@@ -139,46 +141,20 @@ def ended_employment_within_three_months?
       form_object.ended_employment_within_three_months&.yes?
     end
 
+    delegate :kase, :income, :income_payments, :income_benefits, to: :crime_application
```
Comment (timpeat) [inline app/services/decisions/income_decision_tree.rb:144]: Some of these are no longer needed

File: spec/forms/steps/capital/answers_form_spec.rb [hunk 5]
```diff
@@ -12,7 +12,7 @@
 
   let(:attributes) { {} }
 
-  let(:crime_application) { instance_double(CrimeApplication, capital:) }
+  let(:crime_application) { CrimeApplication.new(capital:) }
```
Comment (timpeat) [inline spec/forms/steps/capital/answers_form_spec.rb:15]: don't need this change now

File: spec/validators/application_fulfilment_validator_spec.rb [hunk 6]
```diff
@@ -38,12 +45,19 @@ def to_param
...
+
   let(:employment_status) { [] }
 
   let(:documents) { double(stored: stored_documents) }
   let(:stored_documents) { [] }
 
+  before do
+    allow(capital).to receive(:valid?).with(:submission).and_return(true)
+    allow(kase).to receive(:valid?).with(:submission).and_return(true)
+    allow(kase).to receive(:valid?).with(:submission).and_return(true)
```
Comment (timpeat) [inline spec/validators/application_fulfilment_validator_spec.rb:58]: duplicate

**Review by arthurashman — APPROVED** [review]
Excellent, thank you for running through it with us 🎉

**Pull request merged**